main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala

```diff
@@ -0,0 +1,353 @@
+
+    val getTagMethod = visitor.visitMethod(ACC_PUBLIC + ACC_ABSTRACT, "getTag",
+      s"()${asm.Type.getDescriptor(Constants.stringClass)}", null, null)
+    getTagMethod.visitEnd()
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  /**
+    * Generates an interface for each enum.
+    * Each case of the enum implements this interface. This interface extends `TagInterface`.
+    * @param name name of the class
+    * @param tagInterface the tagInterface
+    * @return byte code representation of the class
+    */
+  def compileEnumInterface(name: List[String], tagInterface: Class[_]): Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(V1_8, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(name), null,
+      asm.Type.getInternalName(Constants.objectClass), Array(asm.Type.getInternalName(tagInterface)))
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  /**
+    * Generates the class for each enum case.
+    * Each class implements the interface of enum.
+    * @param name name of the class
+    * @param enumInterface interfaces for enums
+    * @param tpe descriptor of the `value` field of the class
+    * @param fType Type of the `value` field
+    * @return byte code representation of the class
+    */
+  def compileEnumClass(name: List[String], enumInterface: Class[_], tpe: String, fType: Type): Array[Byte] = {
+    // Initialize the class writer.
+    val visitor = new ClassWriter(ClassWriter.COMPUTE_FRAMES)
+
+    // Initialize the visitor to create a class.
+    visitor.visit(V1_8, ACC_PUBLIC, decorate(name), null, asm.Type.getInternalName(Constants.objectClass),
+      Array(asm.Type.getInternalName(enumInterface)))
+
+    // Generate tag and value fields
+    compileEnumField(visitor, "value", tpe)
+    compileEnumField(visitor, "tag", asm.Type.getDescriptor(Constants.stringClass))
+
+    // Generate the constructor of the class
+    compileEnumConstructor(visitor, name, tpe, fType)
+
+    // Generate the `getValue` method
+    // Picking the right opcode for return
+    val retCode = fType match {
+      case Type.Var(id, kind) =>  throw InternalCompilerException(s"Non-monomorphed type variable '$id in type '$tpe'.")
+      case Type.Bool | Type.Char | Type.Int8 | Type.Int16 | Type.Int32 => IRETURN
+      case Type.Int64 => LRETURN
+      case Type.Float32 => FRETURN
+      case Type.Float64 => DRETURN
+      case _ => ARETURN
+    }
+    compileGetFieldMethod(visitor, name, tpe, "value", "getValue", retCode)
+    // Generate the `getTag` method
+    compileGetFieldMethod(visitor, name, asm.Type.getDescriptor(Constants.stringClass), "tag", "getTag", ARETURN)
+
+    // Generate `hashCode` method
+    compileHashCodeMethod(visitor, name, tpe, fType)
+
+    // Generate `toString` method
+    compileToStringMethod(visitor, name, tpe, fType)
+
+    // Generate `equals` method
+    compileEqualsMethod(visitor, name, tpe, fType)
+
+    visitor.visitEnd()
```

Comment:
I am really happy that you separated out the methods compileToStringMethod etc etc.